x-pack/plugins/aiops/server/routes/queries/fetch_significant_categories.ts

```diff
@@ -0,0 +1,135 @@
+import { LOG_RATE_ANALYSIS_P_VALUE_THRESHOLD } from '../../../common/constants';
+
+import { fetchCategories } from './fetch_categories';
+import { fetchCategoryCounts } from './fetch_category_counts';
+import { getNormalizedScore } from './get_normalized_score';
+
+const getCategoriesTestData = (categories: Category[]): Histogram[] => {
+  const categoriesBaselineTotalCount = getCategoriesTotalCount(categories);
+  return categories.map((d) => ({
+    key: d.key,
+    doc_count: d.count,
+    percentage: d.count / categoriesBaselineTotalCount,
+  }));
+};
+
+const getCategoriesTotalCount = (categories: Category[]): number =>
+  categories.reduce((p, c) => p + c.count, 0);
+
+export const fetchSignificantCategories = async (
+  esClient: ElasticsearchClient,
+  params: AiopsLogRateAnalysisSchema,
+  fieldNames: string[],
+  logger: Logger,
+  // The default value of 1 means no sampling will be used
+  sampleProbability: number = 1,
+  emitError: (m: string) => void,
+  abortSignal?: AbortSignal
+) => {
+  // To make sure we have the same categories for both baseline and deviation,
+  // we do an initial query that spans across baseline start and deviation end.
+  // We could update this to query the exact baseline AND deviation range, but
+  // wanted to avoid the refactor here and it should be good enough for a start.
+  const categoriesOverall = await fetchCategories(
+    esClient,
+    params,
+    fieldNames,
+    params.baselineMin,
+    params.deviationMax,
+    logger,
+    sampleProbability,
+    emitError,
+    abortSignal
+  );
+
+  if (categoriesOverall.length !== fieldNames.length) return [];
+
+  const significantCategories: SignificantTerm[] = [];
+
+  // Using for...of to allow `await` within the loop.
+  for (const [i, fieldName] of fieldNames.entries()) {
+    const categoriesBaseline = await fetchCategoryCounts(
+      esClient,
+      params,
+      fieldName,
+      categoriesOverall[i],
+      params.baselineMin,
+      params.baselineMax,
+      logger,
+      emitError,
+      abortSignal
+    );
+
+    const categoriesDeviation = await fetchCategoryCounts(
+      esClient,
+      params,
+      fieldName,
+      categoriesOverall[i],
+      params.deviationMin,
+      params.deviationMax,
+      logger,
+      emitError,
+      abortSignal
+    );
+
+    const categoriesBaselineTotalCount = getCategoriesTotalCount(categoriesBaseline.categories);
+    const categoriesBaselineTestData = getCategoriesTestData(categoriesBaseline.categories);
+
+    const categoriesDeviationTotalCount = getCategoriesTotalCount(categoriesDeviation.categories);
+    const categoriesDeviationTestData = getCategoriesTestData(categoriesDeviation.categories);
+
+    // Get all unique keys from both arrays
+    const allKeys: string[] = Array.from(
+      new Set([
+        ...categoriesBaselineTestData.map((term) => term.key.toString()),
+        ...categoriesDeviationTestData.map((term) => term.key.toString()),
+      ])
+    ).slice(0, 100);
```

Comment:
taking the first 100 items of an array made by joining two arrays, isn't there a risk it'll contain more items from the first array?

Reply:
Good point! I think that slice was a leftover from copying code or prototyping, I removed it in [635178a](https://github.com/elastic/kibana/pull/165124/commits/635178a68552e39acf70cc832267f7ffe8bf5481).